sdk/metric/batcher/defaultkeys/defaultkeys.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package defaultkeys // import "go.opentelemetry.io/otel/sdk/metric/batcher/defaultkeys"
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/api/core"
+	export "go.opentelemetry.io/otel/sdk/export/metric"
+)
+
+type (
+	Batcher struct {
+		selector export.AggregationSelector
+		lencoder export.LabelEncoder
+		stateful bool
+		dki      dkiMap
+		agg      aggMap
+	}
+
+	aggEntry struct {
+		descriptor *export.Descriptor
+		labels     export.Labels
+		aggregator export.Aggregator
+	}
+
+	dkiMap map[*export.Descriptor]map[core.Key]int
+	aggMap map[string]aggEntry
+
+	producer struct {
+		aggMap   aggMap
+		lencoder export.LabelEncoder
+	}
+)
+
+var _ export.Batcher = &Batcher{}
+var _ export.Producer = &producer{}
+
+func New(selector export.AggregationSelector, lencoder export.LabelEncoder, stateful bool) *Batcher {
+	return &Batcher{
+		selector: selector,
+		lencoder: lencoder,
+		dki:      dkiMap{},
+		agg:      aggMap{},
+		stateful: stateful,
+	}
+}
+
+func (b *Batcher) AggregatorFor(descriptor *export.Descriptor) export.Aggregator {
+	return b.selector.AggregatorFor(descriptor)
+}
+
+func (b *Batcher) Process(_ context.Context, desc *export.Descriptor, labels export.Labels, agg export.Aggregator) error {
+	keys := desc.Keys()
+
+	// Cache the mapping from Descriptor->Key->Index
+	ki, ok := b.dki[desc]
+	if !ok {
+		ki = map[core.Key]int{}
+		b.dki[desc] = ki
+
+		for i, k := range keys {
+			ki[k] = i
+		}
+	}
+
+	// Compute the value list.  Note: Unspecified values become
+	// empty strings.  TODO: pin this down, we have no appropriate
+	// Value constructor.
+	canon := make([]core.KeyValue, len(keys))
+
+	for i, key := range keys {
+		canon[i] = key.String("")
+	}
+
+	// Note also the possibility to speed this computation of
+	// "encoded" via "canon" in the form of a (Descriptor,
+	// LabelSet)->(Labels, Encoded) cache.
+	for _, kv := range labels.Ordered() {
+		pos, ok := ki[kv.Key]
+		if !ok {
+			continue
+		}
+		canon[pos].Value = kv.Value
+	}
+
+	// Compute an encoded lookup key.
+	encoded := b.lencoder.EncodeLabels(canon)
+
+	// Reduce dimensionality.
+	rag, ok := b.agg[encoded]
+	if !ok {
+		// If this Batcher is stateful, create a copy of the
+		// Aggregator for long-term storage.  Otherwise the
+		// Meter implementation will checkpoint the aggregator
+		// again, overwriting the long-lived state.
```

Comment:
the fact that `Aggregator`s need to have ownership (e.g. owned by the `Meter` or owned by the `Batcher`) feels weird to me :/

Reply:
Can you say more about this feeling weird?

The Meter owns the Aggregators that it allocates via AggregatorFor(). During collection, it checkpoints them, at which point the Batcher can read them freely. If the Batcher is stateless, that's the end of the story.

This idea of a stateful Batcher might be the thing that no one really cares about--but the Batcher can't keep a process-lifetime aggregation without owning its own copy. The use-case I've heard for this would be, for example, to let the stdout exporter print process-lifetime sums for counter instruments.